sources/bloodhound/src/main.rs

```diff
@@ -0,0 +1,199 @@
+            }
+
+            if let Ok(checker_data) = serde_json::from_str::<CheckerMetadata>(&metadata) {
+                if checker_data.level <= level {
+                    result.insert(
+                        // Assuming here that we will never have non-ASCII characters in our usage
+                        entry.path().as_os_str().to_string_lossy().to_string(),
+                        checker_data,
+                    );
+                }
+            } else {
+                eprintln!("Unable to parse checker metadata from {:?}", entry.path());
+            }
+        }
+    }
+
+    result
+}
+
+/// Processes the output results from calling an individual checker. Results of
+/// parsing the checker output are added to the overall execution report.
+fn process_checker_results(output: Output, data: CheckerMetadata, report: &mut ReportResults) {
+    if output.status.success() {
+        let check_output = String::from_utf8_lossy(&output.stdout).to_string();
+        if let Ok(checker_result) = serde_json::from_str::<CheckerResult>(&check_output) {
+            report.add_result(data, checker_result);
+            return;
+        }
+    }
+
+    let check_output = String::from_utf8_lossy(&output.stderr).to_string();
+    report.add_result(
+        data,
+        CheckerResult {
+            status: CheckStatus::SKIP,
+            error: check_output,
+        },
+    );
+}
+
+fn get_output(output: &String) -> Result<Box<dyn Write>, Error> {
+    if output.is_empty() {
+        Ok(Box::new(stdout()))
+    } else {
+        File::create(output).map(|f| Box::new(f) as Box<dyn Write>)
+    }
+}
+
+fn main() {
+    let args: Arguments = argh::from_env();
+
+    // Discover all checkers in checks directory
+    let check_dir = PathBuf::from(args.check_dir);
+    if !check_dir.is_dir() {
+        eprintln!("Checker path {:?} is not a directory!", check_dir);
+        std::process::exit(CHECKER_DISCOVERY_ERROR);
+    }
+
+    // Look through the directory and find any checkers, then filter out checks based on input
+    let checkers = find_checkers(&check_dir, args.level);
+    if checkers.is_empty() {
+        eprintln!("No checkers found in {:?}!", check_dir);
+        std::process::exit(CHECKER_DISCOVERY_ERROR);
+    }
+
+    let report_metadata = read_metadata(&check_dir);
+    let mut report = ReportResults::new(args.level, report_metadata);
+
+    // Execute each checker and capture results
+    for (checker, data) in checkers {
+        if let Ok(output) = Command::new(checker).output() {
+            process_checker_results(output, data, &mut report);
+        } else {
+            // Something failed in execution, mark as skipped with message
+            let msg = format!("Error executing {} checker.", data.name);
+            report.add_result(
+                data,
+                CheckerResult {
+                    status: CheckStatus::SKIP,
+                    error: msg,
+                },
+            );
```

Comment:
What would the user do differently in these cases? We know internally that there are different cases that can lead to the current `SKIP` status, but that's not something the end user would know or care about. From their perspective, either bloodhound can tell them if the system is compliant or not, or they need to do some manual verification to make that determination on their own.

The idea of something being only applicable on a given architecture or kernel version is interesting. We don't currently have something like that, but if something were to be introduced in the future, I could see it being handled in a few different ways:

- Have the compilation take in to account the architecture it is being compiled for and include/exclude appropriately
- Default a non-applicable platform/check combination to `PASS` since to the end user, there is no failure and there is nothing to manually check
- Introduce another test state if there is a need to report this non-applicable state

I think the second one is probably the UX we would want, but until we have one of those cases I don't think we have the full context to make that decision right now. So I would rather defer until there is a need before trying to proactively handle it.

Reply:
>What would the user do differently in these cases?

My main concern is really that someone will read "SKIP" and decide that the check was skipped because it didn't apply, not because they actually need to dig in and do additional work. But the error messages should help clarify the distinction between these two cases.